Added ability to consecutive calls for WhenCalled.Do method

bartoszgolek opened the pull request.

dtchepak commented:
Thanks for the PR!

I have a couple of questions/comments about this feature.

* What order are these calls meant to be replayed? (At present they replay in the opposite order to the way they were queued)
* What should happen when the end of the queue is reached? At present it throws an exception. With `.Returns(x,y,z)`, [NSubstitute will keep returning `z` for the third and later calls](https://github.com/nsubstitute/NSubstitute/blob/0e70eb36eff9b017a66e3d73b4390c0278042bf1/Source/NSubstitute.Acceptance.Specs/ReturningResults.cs#L37). Keeping this behaviour would be consistent, but maybe it is worth a differently named call that will only empty the stack then do nothing after that (`.DoOnly(...)`? Not sure about a good name)?
* If we can define a good, consistent behaviour for this then it can probably go straight into `WhenCalled.cs` rather than an extension.
* Will need some acceptance tests (not unit tests unless you feel a need for them). There is a `WhenCalledDo` spec already.

Here's an example test I was looking at:

```
        [Test]
        public void Consecutive_calls()
        {
            var ints = new List<int>();
            _something
                .When(x => x.Count())
                .Do(x => ints.Add(1), x => ints.Add(2), x => ints.Add(3));

            _something.Count(); // 1
            _something.Count(); // 2
            _something.Count(); // 3
            _something.Count(); // 4th call??

            Assert.That(ints, Is.EqualTo(new[] { 1, 2, 3 })); //??
        }
```

bartoszgolek commented:
Hi,

Thanks for the reply.
I've added acceptance tests as You wished.

I've also added, 'do nothing' behavior when no more calls defined.
Whats still wondering me is, two similar calls give different effects:
`_something.When(x => x.Count()).Do(new Action<CallInfo>[] { x => ints.Add(1) });` will do nothing for second call
`_something.When(x => x.Count()).Do(x => ints.Add(1));` will call `ints.Add` for every call.
Maybe there should be difference in MethodName too?

dtchepak commented:
Yes I agree it is confusing.

One option rather than 'do nothing' is to repeat the final call. (So the assertion in the test I sent would become `Assert.That(ints, Is.EqualTo(new[] { 1, 2, 3, 3 }));`)

Another option is to have a different method to handle cases where we only want to run a fixed number of callbacks, but I'm not sure what to name it.

// ping @AnthonyEgerton

bartoszgolek commented:
Ok, I checked how Returns works and it repeat the last call, so Do should work in the same way. I made it cohesiv. Also, I added documentation tags and moved method to WhenCalled.cs

dtchepak commented:
Thank you for the revisions.

One difference between When..Do and Returns is that we can override Returns. For example:
```
a.Foo().Returns(1); // Foo() now returns 1
a.Foo().Returns(2); // Foo() now returns 2.
```

We can't do the same with When..Do -- once a callback is set it will always run. So with that in mind, would `a.When(x => x.Foo()).Do(x, y, z)` do all three calls whenever `Foo()` is called, or call each one in sequence?

It would be nice to be able to use the code you've added to support temporary callbacks, but I think this would require different naming.

With this in mind, it would be great if we could come up with a name to describe queueing up these callbacks so differentiate it from the standard `When..Do` behaviour.

What do you think of something like `a.When(x => x.Foo()).RunQueue(x, y, z)`?
(`DoQueue`, `RunQueue`, `DoOnly`, ... ?)

bartoszgolek commented:
Maybe, if I explain why I need this it will be easier.

In my case I have web service that i have to call in my code. The customer, said that in case of fail, call should be repeated till success or reach count limit. Service method is async, so I cant use Returns do to void result of service method.

That's the case.

bartoszgolek commented:
Hi,

I'm still wondering how to name this behavior.
Is it necessary to have Do as collection of calls? What is the purpose fo adding many Do callbacks instead of one doing all things thogether?

This behavior makes this complicated.
First of all, i was thinging that do works in the same way as Returns. I think this is inconsistent.
I dont want to turn NSubstitute upside down, but im trying to understand the feature.

Best regrds
Bartosz

dtchepak commented:
Hi Bartosz, 
Sorry for the delay. I hadn't forgotten this issue, I just need to find some time to experiment with some different ideas around it. I'm thinking of trying to pull out the queueing behaviour so we can keep the existing syntax, but expose a convenient way of queuing callbacks.

> What is the purpose fo adding many Do callbacks instead of one doing all things thogether?

Do you mean why do `.Do` not override previously configured callbacks? I don't remember the rational for the original decision, but I'm guessing it was to support queuing up different callbacks, say one in `[SetUp]` relevant to all the tests, and a more specific one in one particular test. This is a bit different to the `Returns` case - we only ever want to return one value at a time, but we may need to run multiple actions at a time.

I have a feeling this doesn't answer your question correctly, so please post clarifications if that is the case.

bartoszgolek commented:
Currently it explains exactly what I'm asking for.

What I understand:
1. When.Do is to call additional actions during the test flow
2. Returns is to configure substitution behavior

What we need in that case is to configure void Method behavior.
After Your explanantion i've got an idea.

Any void method can do only two things:
 - finish wihout problems
 - throw an exception

Maybe it is a good way to add When.Throw behavior for this scenario.
It can looks like:
```
a.When(x => x.Foo()).Throw(() => new Exception(), () => new ArgumentNullException());
```

It should work like:

```
		[Test]
		public void will_repeat_last_call_when_no_more_consecutive_calls_exists()
		{
			var ints = new List<int>();
			var i = 0;
			_something
				.When(x => x.Count())
				.Do(x => ints.Add(++i))
				.Throws(() => new Exception(), () => new ArgumentNullException());

			Assert.Throws<Exception>(() => _something.Count()); // 1 + Throw
			Assert.Throws<ArgumentNullException>(() => _something.Count()); //2 + Throw
			Assert.DoesNotThrow(() => _something.Count()); // 3

			Assert.That(ints, Is.EqualTo(new[] { 1, 2, 3}));
		}
```

In that case we can leave When.Do, as it is, and add functionality to Throw exception from void calls.
What do You think?

dtchepak commented:
I like the direction you're taking, but in that particular example it looks to me like it will `Do` the callback, then throw `Exception` on the next call, then throw `ArgumentNullException`.

What if we expose the queue as a callback-builder kind of thing? We could add a `Do` overload that accepts a `Callbacks` argument.

        [Test]
        public void will_repeat_last_call_when_no_more_consecutive_calls_exists()
        {
            var ints = new List<int>();
            var i = 0;
            _something
                .When(x => x.Count())
                .Do(Callbacks.Throw<Exception>()
                             .ThenThrow<ArgumentNullException>()
                             .Always(x => i++));

            Assert.Throws<Exception>(() => _something.Count()); // 1 + Throw
            Assert.Throws<ArgumentNullException>(() => _something.Count()); //2 + Throw
            Assert.DoesNotThrow(() => _something.Count()); // 3

            Assert.That(ints, Is.EqualTo(new[] { 1, 2, 3}));
        }

I'm thinking we could leave the `Do` syntax pretty much unchanged in that case, but support all kinds of callback configurations via the builder. For example:

    .Do(Callbacks.First(x => list.Add("first call"))
                 .Then(x => list.Add("second call"))
                 .ThenThrow<Exception>()
                 .Then(x => list.Add("fourth call"))

Implementation-wise, `Do(CallBacks c)` could be implemented in terms of the original `Do`, something like:

    Do(x => c.RunNext(x))

We could possibly adapt this `Callbacks` type to also work with `Returns`, which might help when trying stub a mix of return values and exceptions from a call.

Thoughts?

bartoszgolek commented:
Looks impressive and covers all scenarios!!
Lets do this!

Tommorow I will try with some Aceptance tests and implementations. For now only for Do, Returns can be done in another pull request.
I think it's a good idea to close this PR and create new one with detailed explanation.
